ADR/0027-availability-probe-framework.md

```diff
@@ -0,0 +1,102 @@
+# 27. Availability Probe Framework
+
+* Date 2023-07-06
+
+## Status
+
+Accepted
+
+## Context
+
+As a product team, we want a composite system availability metric for both Production
+and Staging environments, so that we can make roadmap decisions with easy-to-understand
+feedback about the reliability of the system in mind.
+  
+This metric should tell us the overall up time of RHTAP with respect to services under
+the control of Red Hat.
+
+RHTAP is defined to be available at a given moment if all of its components are
+reporting to be available at that given moment.
+
+A component is defined to be available at a given moment if all of its availability
+probes are reporting to be available at that given moment.
+
+A convention is required for providing the availability of a probe.
+
+Once this is in place, those metrics can be aggregated in order to report the overall
+availability of RHTAP.
+
+## Decision
+
+Probes' availability will be provided as a Prometheus metric. The metric will contain
+labels to allow differentiating between the different probes. When exported out of the
+cluster of origin, additional labels will be attached to the metric to mark the cluster
+in which the metric was generated.
+
+### Delivery Method
+
+The availability metric will be computed for each component based on the exit status of
+the latest execution of a CronJob evaluating the component's availability. The different
+teams will provide the CronJobs' implementation, while the RHTAP Observability team will
+provide the means to translate the jobs' results to the required metric.
+
+It is up for each team to define what it means for its component(s) to be available.
+Fundamentally, a component should be reported as available as long as it's capable of
+providing the service it aims to provide, and unavailable otherwise.
+
+Each team will define CronJobs that will test the availability of their components.
+The Job started by each CronJob will terminate successfully if the test complete
+successfully, and will terminate with an error in case the test fails.
+
+Kube-state-metrics generates Jobs and CronJob metrics that can be processed using
+Prometheus recording rules in order to generate RHTAP's availability metric.
+
+The RHTAP Observability team will construct the necessary rules.
```

Comment:
Do we need to call out a specific team here (and in other places in the document if I missed them)? I am not sure if we really call out teams in ADRs, instead aggregations/associations seem to be made based on service boundaries. I know that there isn't a service boundary for o11y, but is there some better association?

Reply:
Point taken. I'll try to generalize this throughout the document in a way that will convey to the teams that they only need to worry about the cronjob without calling out who's going to take care of the rest.